src/realm/sync/network/default_socket.cpp

```diff
@@ -491,70 +492,69 @@ void DefaultSocketProvider::start()
-        if (m_thread.joinable()) {
-            lock.unlock();
-            m_thread.join();
-            lock.lock();
-        }
+    if (m_state == State::Stopping) {
+        state_wait_for(lock, State::Stopped);
     }
 
     m_logger_ptr->trace("Default event loop: start()");
     REALM_ASSERT(m_state == State::Stopped);
-    do_state_update(State::Starting);
+    do_state_update(lock, State::Starting);
     m_thread = std::thread{&DefaultSocketProvider::event_loop, this};
-    lock.unlock();
     // Wait for the thread to start before continuing
-    state_wait_for(State::Started);
+    state_wait_for(lock, State::Running);
 }
 
 void DefaultSocketProvider::event_loop()
 {
     m_logger_ptr->trace("Default event loop: thread running");
-    std::unique_lock<std::mutex> lock(m_mutex);
-    // The thread has started - don't return early due to stop() until after did_create_thread()
-    if (m_state == State::Starting) {
-        do_state_update(State::Started);
-    }
-    lock.unlock();
-
     auto will_destroy_thread = util::make_scope_exit([&]() noexcept {
         m_logger_ptr->trace("Default event loop: thread exiting");
         if (g_binding_callback_thread_observer)
             g_binding_callback_thread_observer->will_destroy_thread();
 
-        std::lock_guard<std::mutex> lock(m_mutex);
+        std::unique_lock<std::mutex> lock(m_mutex);
         // Did we get here due to an unhandled exception?
         if (m_state != State::Stopping) {
             m_logger_ptr->error("Default event loop: thread exited unexpectedly");
         }
-        do_state_update(State::Stopped);
+        m_state = State::Stopped;
+        std::notify_all_at_thread_exit(m_state_cv, std::move(lock));
     });
 
     if (g_binding_callback_thread_observer)
         g_binding_callback_thread_observer->did_create_thread();
 
-    lock.lock();
-    if (m_state != State::Started) {
-        // Stop has already been requested - exit early
-        lock.unlock(); // make sure the mutex is unloaded before exiting
-        return;        // early return
+    {
+        std::lock_guard<std::mutex> lock(m_mutex);
+        REALM_ASSERT(m_state == State::Starting);
     }
-    do_state_update(State::Running);
-    lock.unlock();
 
-    m_logger_ptr->trace("Default event loop: service run");
+    // We update the state to Running from inside the event loop so that start() is blocked until
+    // the event loop is actually ready to receive work.
+    m_service.post([this](Status status) {
+        std::unique_lock<std::mutex> lock(m_mutex);
+        if (status == ErrorCodes::OperationAborted) {
+            REALM_ASSERT(m_state == State::Stopping || m_state == State::Stopped);
+            return;
+        }
+
+        REALM_ASSERT(status.is_ok());
+        m_logger_ptr->trace("Default event loop: service run");
+        REALM_ASSERT(m_state == State::Starting);
+        do_state_update(lock, State::Running);
+    });
+
     try {
```

Comment:
You might want to add a check here to confirm the state is still running before it calls `m_service.run()` - if stop() was called before this, you _would_ hit the REALM_ASSERT() in the post function handler.